File: verify_answers.py
"""公式解答PDFから解答を抽出し JSON に書き出す（verify_answers.js が読む）"""
import pypdfium2 as pdfium
import json, re, os

base = os.path.dirname(os.path.abspath(__file__))

# 年度キー -> 公式解答PDF
PDFS = {
    'r1kami':       '20190602_co_second_a01.pdf',
    'r1shimo':      '20191006_co_second_a01.pdf',
    'r2shimo':      '20201004_co_second_a01.pdf',
    'r2shimo_gogo': '20201004_co_second_a02.pdf',
    'r3kami':       '20210530_co_second_a01.pdf',
    'r3kami_gogo':  '20210530_co_second_a02.pdf',
    'r3shimo':      '220211024_co_second_a01.pdf',
    'r3shimo_gogo': '220211024_co_second_a02.pdf',
    'r4kami':       '20220529_co_second_a01.pdf',
    'r4kami_gogo':  '20220529_co_second_a02.pdf',
    'r4shimo':      '20221030_co_second_a01.pdf',
    'r4shimo_gogo': '20221030_co_second_a02.pdf',
    'r5kami':       '20230528_co_second_a01.pdf',
    'r5shimo':      '20231029_co_second_a01.pdf',
    'r5shimo_gogo': '20231029_co_second_a02.pdf',
    'r6kami':       '20240526_co_second_a01.pdf',
    'r6shimo':      '20241027_co_second_a01.pdf',
    'r7kami':       '20250525_co_second_a01.pdf',
    'r7shimo':      '20251026_co_second_a01.pdf',
    'r8kami':       '20260524_co_second_a01.pdf',
}

IDX = {'イ': 0, 'ロ': 1, 'ハ': 2, 'ニ': 3}

result = {}
for key, fname in PDFS.items():
    path = os.path.join(base, 'list', fname)
    if not os.path.exists(path):
        print(f'{key}: PDFなし ({fname})')
        continue
    pdf = pdfium.PdfDocument(path)
    text = pdf[0].get_textpage().get_text_range()
    # 「番号 解答」のペアを全て拾う
    pairs = re.findall(r'(\d{1,2})\s*([イロハニ])', text)
    answers = {}
    for num, ch in pairs:
        n = int(num)
        if 1 <= n <= 50:
            answers[n] = IDX[ch]
    missing = [n for n in range(1, 51) if n not in answers]
    if missing:
        print(f'{key}: 抽出できなかった問 -> {missing}')
    result[key] = {str(n): answers[n] for n in sorted(answers)}
    print(f'{key}: {len(answers)}問 抽出')

out = os.path.join(base, 'official_answers.json')
with open(out, 'w', encoding='utf-8') as f:
    json.dump(result, f, ensure_ascii=False, indent=1)
print('\n->', out)

File: verify_answers.js
// official_answers.json と各年度の *_data.js の answer を照合する
const fs = require('fs');
const path = require('path');

const base = __dirname;
const official = JSON.parse(fs.readFileSync(path.join(base, 'official_answers.json'), 'utf8'));

const KEYS = {
  r1kami: 'R1KAMI', r1shimo: 'R1SHIMO',
  r2shimo: 'R2SHIMO', r2shimo_gogo: 'R2SHIMO_GOGO',
  r3kami: 'R3KAMI', r3kami_gogo: 'R3KAMI_GOGO',
  r3shimo: 'R3SHIMO', r3shimo_gogo: 'R3SHIMO_GOGO',
  r4kami: 'R4KAMI', r4kami_gogo: 'R4KAMI_GOGO',
  r4shimo: 'R4SHIMO', r4shimo_gogo: 'R4SHIMO_GOGO',
  r5kami: 'R5KAMI', r5shimo: 'R5SHIMO', r5shimo_gogo: 'R5SHIMO_GOGO',
  r6kami: 'R6KAMI', r6shimo: 'R6SHIMO',
  r7kami: 'R7KAMI', r7shimo: 'R7SHIMO',
  r8kami: 'R8KAMI',
};

const L = ['イ', 'ロ', 'ハ', 'ニ'];
let totalBad = 0;

console.log('年度'.padEnd(15) + ' 不一致  内訳');
console.log('-'.repeat(70));

for (const [file, key] of Object.entries(KEYS)) {
  const p = './' + file + '_data.js';
  global.window = {};
  delete require.cache[require.resolve(p)];
  require(p);
  const qs = window[key].questions;
  const off = official[file];
  if (!off) { console.log(file.padEnd(15) + ' 公式解答なし'); continue; }

  const bad = [];
  for (const q of qs) {
    const exp = off[String(q.id)];
    if (exp === undefined) { bad.push(`問${q.id}:公式なし`); continue; }
    if (q.answer !== exp) bad.push(`問${q.id}(data:${L[q.answer]}/公式:${L[exp]})`);
  }
  totalBad += bad.length;
  const head = file.padEnd(15) + String(bad.length).padStart(5) + '  ';
  console.log(head + (bad.length ? bad.join(' ') : 'OK'));
}

console.log('-'.repeat(70));
console.log(totalBad === 0 ? '全年度 全50問 解答一致' : `解答の不一致 合計 ${totalBad} 件`);
